### PR TITLE
ScatterPlots: handle xOverrideTime and apply color scales

### PR DESCRIPTION
**Pull request by danielgavrilov**
Also adds more tests around this. Keeps all behaviour as it was before, haven't yet implemented mixed-time-type table joins...

### CODE REVIEW

File: coreTable/OwidTable.test.ts [hunk 1]
```diff
@@ -496,62 +496,76 @@ describe("time domain", () => {
...
-                    year: 2005,
-                }),
-            ])
-        )
-    })
+    function applyTolerance(table: OwidTable) {
+        return table.interpolateColumnWithTolerance("gdp", 1)
+    }
+
+    // Applying the tolerance twice to ensure operation is idempotent.
```
Comment (breck7) [inline coreTable/OwidTable.test.ts:517]: Ah, nice catch! Didn't think of that.

File: coreTable/CoreTable.ts [hunk 2]
```diff
@@ -1061,6 +1061,34 @@ export class CoreTable<
...
+            )
+        }
+
+        const existingColumn = this.get(slug)
+        if (!existingColumn) {
+            throw new Error(`Cannot duplicate column: '${slug}' does not exist`)
+        }
+
+        return this.transform(
+            {
```
Comment (breck7) [inline coreTable/CoreTable.ts:1077]: I like it. This is maybe the pattern to use for the other transforms (just building a new column store then passing that to transform)

File: coreTable/CoreTableUtils.ts [hunk 3]
```diff
@@ -227,7 +228,8 @@ export const makeKeyFn = (
     columnStore: CoreColumnStore,
     columnSlugs: ColumnSlug[]
 ) => (rowIndex: number) =>
-    columnSlugs.map((slug) => columnStore[slug][rowIndex].toString()).join(" ")
+    // toString() handles `undefined` and `null` values, which can be in the table.
```
Comment (breck7) [inline coreTable/CoreTableUtils.ts:231]: Makes sense. Relatedly I think it'd be ideal to keep undefineds and nulls out of the table at some point, but haven't nailed that down yet (and not entirely sure it's important).

**Review by breck7 — APPROVED** [review]
LGTM but I haven't in my head made the connection yet as to the downstream fix on scatters this has (I know it's something to do with date joins, but forget the exacts).

Is there a before and after photo? Or is this a building step toward the fix?

**Comment by danielgavrilov** [comment]
Ok this now handles the legacy `dimension.targetYear`. We had renamed it to `targetTime` but the database config is still under `targetYear` so we'll have to stick to it for now.

The ScatterPlot seems to receive the right table now, but still not fixed, will do that in a separate PR.

**Comment by danielgavrilov** [comment]
Implemented some ScatterPlot fixes:
- [Scatter: handle xOverrideTime](https://www.notion.so/owid/Scatter-handle-xOverrideTime-6fdb71161d2a478aafe19367c6dff58d)
- [Scatter: fix color scales](https://www.notion.so/owid/Scatter-fix-color-scales-532323e7d0084e229789970c559d7235)

There are still others to go, keeping track in [Scatter: issues & broken charts](https://www.notion.so/owid/Scatter-issues-broken-charts-16577b3c1ea449e4a93b9da815fd86be).

Since this introduces some changes in legacy conversion & other fixes, I think we can merge it now, and fix the other ScatterPlot issues separately.

File: adminSite/server/testPageRouter.tsx [hunk 4]
```diff
@@ -97,7 +103,9 @@ function EmbedTestPage(props: EmbedTestPageProps) {
                 </div>
                 {props.charts.map((chart) => (
                     <div key={chart.slug} className="row">
-                        <div className="chart-id">{chart.id}</div>
+                        <div className="chart-id">
+                            <a href={`?ids=${chart.id}`}>{chart.id}</a>
```
Comment (breck7) [inline adminSite/server/testPageRouter.tsx:107]: Good idea!

**Pull request merged**